File: python_implementation/vfd_sh1106_adapter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
VFD_SH1106_Adapter.py
CarRadio SH1106 Display Adapter (Python Version)

This adapter converts VFD commands to SH1106 OLED display commands using the luma.oled library
Optimized for Raspberry Pi deployment
"""

import time
import logging
from PIL import Image, ImageDraw, ImageFont
from luma.core.interface.serial import i2c
from luma.oled.device import sh1106

class VFD_SH1106_Adapter:
    """
    This class implements a similar interface to the original VFD class
    but uses the luma.oled library to drive an SH1106 OLED display
    """
    
    # Constants from original VFD class
    VFD_OUTLINE = 0x14
    VFD_CLEAR_AREA = 0x12
    VFD_SET_AREA = 0x11
    VFD_SET_CURSOR = 0x10
    VFD_SET_WRITEMODE = 0x1A
    
    SCROLL_BAR_WIDTH = 3
    
    # Font types
    FONT_MINI = 0
    FONT_5x7 = 1
    FONT_10x14 = 2
    
    def __init__(self):
        """Initialize the adapter"""
        self._is_setup = False
        self._cursor_x = 0
        self._cursor_y = 0
        self._current_font = self.FONT_MINI
        self._i2c_address = 0x3C  # Default I2C address for most SH1106 displays
        self._device = None
        self._image = None
        self._draw = None
        self._font = None
        self._font_size = 8  # Default font size
        
    def begin(self, path, address=0x3C):
        """
        Initialize the display with I2C
        
        Args:
            path (str): I2C device path (e.g., "/dev/i2c-1")
            address (int): I2C device address (default: 0x3C)
            
        Returns:
            bool: True if initialization was successful, False otherwise
        """
        try:
            # Extract I2C bus number from path
            bus_num = int(path.split('-')[-1])
            
            # Initialize I2C interface
            serial_interface = i2c(port=bus_num, address=address)
            
            # Initialize SH1106 device
            self._device = sh1106(serial_interface, width=128, height=64, rotate=0)
            
            # Create image buffer
            self._image = Image.new('1', (self._device.width, self._device.height))
            self._draw = ImageDraw.Draw(self._image)
            
            # Set default font
            self.set_font(self.FONT_MINI)
            
            # Clear the display
            self.clear_screen()
            
            # Set full brightness
            self.set_brightness(7)
            
            self._is_setup = True
            return True
        except Exception as e:
            logging.error(f"Error initializing display: {e}")
            return False
    
    def stop(self):
        """Clean up resources"""
        if self._device:
            self._device.cleanup()
            self._device = None
        self._is_setup = False
    
    def reset(self):
        """Reset the display"""
        if not self._is_setup:
            return False
        
        self._device.clear()
        self._image = Image.new('1', (self._device.width, self._device.height))
        self._draw = ImageDraw.Draw(self._image)
        self._cursor_x = 0
        self._cursor_y = 0
        return True
    
    def set_brightness(self, level):
        """
        Set the display brightness
        
        Args:
            level (int): Brightness level (0-7), where 0 is off and 7 is maximum
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        # Convert 0-7 scale to 0-255
        brightness = min(255, int(level * 255 / 7))
        self._device.contrast(brightness)
        return True
    
    def set_power_on(self, set_on):
        """
        Turn the display on or off
        
        Args:
            set_on (bool): True to turn on, False to turn off
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        if set_on:
            self._device.show()
        else:
            self._device.hide()
        return True
    
    def set_cursor(self, x, y):
        """
        Set the cursor position
        
        Args:
            x (int): X coordinate
            y (int): Y coordinate
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        self._cursor_x = x
        self._cursor_y = y
        return True
    
    def set_font(self, font):
        """
        Set the font
        
        Args:
            font (int): Font type (FONT_MINI, FONT_5x7, or FONT_10x14)
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        self._current_font = font
        
        # Update font based on selected type
        try:
            if font == self.FONT_MINI:
                self._font_size = 8
                self._font = ImageFont.truetype("DejaVuSansMono.ttf", self._font_size)
            elif font == self.FONT_5x7:
                self._font_size = 12
                self._font = ImageFont.truetype("DejaVuSansMono.ttf", self._font_size)
            elif font == self.FONT_10x14:
                self._font_size = 16
                self._font = ImageFont.truetype("DejaVuSansMono.ttf", self._font_size)
            else:
                # Default to mini font
                self._font_size = 8
                self._font = ImageFont.truetype("DejaVuSansMono.ttf", self._font_size)
        except IOError:
            # If font file not found, use default bitmap font
            if font == self.FONT_MINI:
                self._font_size = 8
                self._font = ImageFont.load_default()
            elif font == self.FONT_5x7:
                self._font_size = 12
                self._font = ImageFont.load_default()
            elif font == self.FONT_10x14:
                self._font_size = 16
                self._font = ImageFont.load_default()
            else:
                # Default to mini font
                self._font_size = 8
                self._font = ImageFont.load_default()
                
        return True
    
    def clear_screen(self):
        """
        Clear the screen
        
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        self._draw.rectangle((0, 0, self._device.width-1, self._device.height-1), fill=0)
        self._device.display(self._image)
        return True
    
    def draw_scroll_bar(self, top, bar_height, starting_offset):
        """
        Draw a scroll bar
        
        Args:
            top (int): Top position of the scroll bar
            bar_height (float): Height of the scroll bar as a proportion (0.0-1.0)
            starting_offset (float): Starting offset as a proportion (0.0-1.0)
            
        Returns:
            None
        """
        if not self._is_setup:
            return
        
        # Calculate scroll bar dimensions
        display_height = self._device.height
        scroll_bar_x = self._device.width - self.SCROLL_BAR_WIDTH
        
        # Draw scroll bar background
        self._draw.rectangle(
            (scroll_bar_x, 0, self._device.width-1, display_height-1),
            fill=0
        )
        
        # Calculate scroll bar position and size
        bar_top = int(starting_offset * display_height)
        bar_height_px = max(4, int(bar_height * display_height))
        
        # Draw scroll bar
        self._draw.rectangle(
            (scroll_bar_x, bar_top, self._device.width-1, bar_top + bar_height_px - 1),
            fill=1
        )
        
        # Update display
        self._device.display(self._image)
    
    def write(self, text):
        """
        Write text at the current cursor position
        
        Args:
            text (str): Text to write
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        self._draw.text((self._cursor_x, self._cursor_y), text, font=self._font, fill=1)
        self._device.display(self._image)
        return True
    
    def write_packet(self, data, length, wait_usec=50):
        """
        Write a packet of data
        
        Args:
            data (bytes): Data to write
            length (int): Length of data
            wait_usec (int): Wait time in microseconds (not used in this implementation)
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup or length == 0:
            return False
        
        # Check if this is a command
        if data[0] == self.VFD_SET_CURSOR and length >= 3:
            # Set cursor command
            x = data[1]
            y = data[2]
            self.set_cursor(x, y)
        elif data[0] == self.VFD_CLEAR_AREA and length >= 5:
            # Clear area command
            x1 = data[1]
            y1 = data[2]
            x2 = data[3]
            y2 = data[4]
            self._draw.rectangle((x1, y1, x2, y2), fill=0)
            self._device.display(self._image)
        else:
            # Treat as text data
            try:
                text = data[0:length].decode('utf-8')
                self.write(text)
            except UnicodeDecodeError:
                # If not valid UTF-8, just write each byte as a character
                for i in range(length):
                    self.write(chr(data[i]))
        
        return True
    
    def print_packet(self, fmt, *args):
        """
        Print a formatted packet
        
        Args:
            fmt (str): Format string
            *args: Arguments for the format string
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        text = fmt % args
        return self.write(text)
    
    def print_lines(self, y, step, lines, first_line, max_lines, font=FONT_MINI):
        """
        Print multiple lines of text
        
        Args:
            y (int): Starting Y position
            step (int): Vertical step between lines
            lines (list): List of strings to print
            first_line (int): Index of the first line to print
            max_lines (int): Maximum number of lines to print
            font (int): Font to use
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        # Set the font
        self.set_font(font)
        
        # Calculate how many lines we can display
        num_lines = len(lines)
        display_lines = min(max_lines, num_lines - first_line)
        
        if display_lines <= 0:
            return True  # Nothing to display
        
        # Display each line
        for i in range(display_lines):
            line_idx = first_line + i
            if line_idx < num_lines:
                self.set_cursor(0, y + i * step)
                self.write(lines[line_idx])
        
        return True
    
    def print_rows(self, y, step, columns, first_line, max_lines, x_offset=0, font=FONT_MINI):
        """
        Print rows with multiple columns
        
        Args:
            y (int): Starting Y position
            step (int): Vertical step between rows
            columns (list): List of lists, where each inner list contains strings for each column
            first_line (int): Index of the first row to print
            max_lines (int): Maximum number of rows to print
            x_offset (int): Horizontal offset for the first column
            font (int): Font to use
            
        Returns:
            bool: True if successful, False otherwise
        """
        if not self._is_setup:
            return False
        
        # Set the font
        self.set_font(font)
        
        # Calculate how many rows we can display
        num_rows = len(columns)
        display_rows = min(max_lines, num_rows - first_line)
        
        if display_rows <= 0:
            return True  # Nothing to display
        
        # Calculate column widths
        col_widths = []
        max_cols = 0
        
        for row in columns:
            max_cols = max(max_cols, len(row))
        
        if max_cols == 0:
            return True  # No columns to display
        
        # Find the maximum width for each column
        for col_idx in range(max_cols):
            max_width = 0
            for row in columns:
                if col_idx < len(row):
                    # Estimate text width based on character count and font size
                    # This is a rough approximation
                    text_width = len(row[col_idx]) * (self._font_size // 2)
                    max_width = max(max_width, text_width)
            col_widths.append(max_width + 5)  # Add some padding
        
        # Display each row
        for i in range(display_rows):
            row_idx = first_line + i
            if row_idx < num_rows:
                row = columns[row_idx]
                y_pos = y + i * step
                
                # Display each column in the row
                x_pos = x_offset
                for col_idx, col_text in enumerate(row):
                    self.set_cursor(x_pos, y_pos)
                    self.write(col_text)
                    
                    # Move to the next column position
                    if col_idx < len(col_widths):
                        x_pos += col_widths[col_idx]
        
        return True
    
    def width(self):
        """
        Get the display width
        
        Returns:
            int: Display width in pixels
        """
        return 128 if self._device else 0
    
    def height(self):
        """
        Get the display height
        
        Returns:
            int: Display height in pixels
        """
        return 64 if self._device else 0
    
    @staticmethod
    def truncate(text, width):
        """
        Truncate text to fit within a specified width
        
        Args:
            text (str): Text to truncate
            width (int): Maximum width
            
        Returns:
            str: Truncated text
        """
        if len(text) <= width:
            return text
        return text[:width-3] + "..."

File: VFD_SH1106_Adapter.cpp
//
//  VFD_SH1106_Adapter.cpp
//  carradio_v2_sh1106
//
//  Created on 2/26/2025
//  This adapter converts VFD commands to SH1106 OLED display commands using the u8g2 library
//  Optimized for Raspberry Pi deployment
//

#include "VFD_SH1106_Adapter.hpp"
#include <stdarg.h>

VFD_SH1106_Adapter::VFD_SH1106_Adapter() {
    _isSetup = false;
    _cursorX = 0;
    _cursorY = 0;
    _currentFont = FONT_MINI;
    _u8g2 = nullptr;
    _i2cAddress = 0x3C; // Default I2C address for most SH1106 displays
}

VFD_SH1106_Adapter::~VFD_SH1106_Adapter() {
    stop();
}

bool VFD_SH1106_Adapter::begin(const char* path, speed_t speed) {
    int error = 0;
    return begin(path, speed, error);
}

bool VFD_SH1106_Adapter::begin(const char* path, speed_t speed, int &error) {
    // For compatibility with the original VFD class
    // On Raspberry Pi, we'll use the I2C address version instead
    return begin(path, _i2cAddress);
}

bool VFD_SH1106_Adapter::begin(const char* path, uint8_t i2cAddress) {
    // Raspberry Pi specific initialization
    _i2cAddress = i2cAddress;
    
    // Create the U8G2 object for SH1106 with I2C
    // U8G2_R0 means no rotation, display is drawn normal
    _u8g2 = new U8G2_SH1106_128X64_NONAME_F_HW_I2C(U8G2_R0);
    
    if (_u8g2 == nullptr) {
        return false;
    }
    
    // Set I2C address if different from default
    if (_i2cAddress != 0x3C) {
        _u8g2->setI2CAddress(_i2cAddress * 2); // U8g2 requires the address to be shifted
    }
    
    // Initialize the display
    _u8g2->begin();
    _u8g2->clearBuffer();
    _u8g2->sendBuffer();
    
    // Set default font
    setFont(FONT_MINI);
    
    _isSetup = true;
    return _isSetup;
}

void VFD_SH1106_Adapter::stop() {
    if (_isSetup && _u8g2 != nullptr) {
        delete _u8g2;
        _u8g2 = nullptr;
    }
    _isSetup = false;
}

bool VFD_SH1106_Adapter::reset() {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    _u8g2->initDisplay();
    _u8g2->clearBuffer();
    _u8g2->sendBuffer();
    return true;
}

bool VFD_SH1106_Adapter::setBrightness(uint8_t level) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    // Map 0-7 to 0-255
    uint8_t contrast = (level * 255) / 7;
    _u8g2->setContrast(contrast);
    return true;
}

bool VFD_SH1106_Adapter::setPowerOn(bool setOn) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    _u8g2->setPowerSave(!setOn);
    return true;
}

bool VFD_SH1106_Adapter::setCursor(uint8_t x, uint8_t y) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    _cursorX = x;
    _cursorY = y;
    return true;
}

bool VFD_SH1106_Adapter::setFont(font_t font) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    _currentFont = font;
    updateFont();
    return true;
}

void VFD_SH1106_Adapter::updateFont() {
    switch (_currentFont) {
        case FONT_MINI:
            _u8g2->setFont(u8g2_font_5x8_tr); // Small font
            break;
        case FONT_5x7:
            _u8g2->setFont(u8g2_font_6x10_tr); // Medium font
            break;
        case FONT_10x14:
            _u8g2->setFont(u8g2_font_10x20_tr); // Large font
            break;
    }
}

bool VFD_SH1106_Adapter::clearScreen() {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    _u8g2->clearBuffer();
    _u8g2->sendBuffer();
    return true;
}

void VFD_SH1106_Adapter::drawScrollBar(uint8_t topbox, float bar_height, float starting_offset) {
    if (!_isSetup || _u8g2 == nullptr)
        return;
    
    uint8_t rightbox = width() - 1;
    uint8_t leftbox = rightbox - scroll_bar_width + 1;
    uint8_t bottombox = 63;
    uint8_t scroll_height = bottombox - topbox - 2;
    uint8_t bar_size = ceil(scroll_height * bar_height);
    uint8_t offset = ((scroll_height - bar_size) * starting_offset) + topbox + 1;
    
    // Draw outline
    _u8g2->drawFrame(leftbox, topbox, scroll_bar_width, bottombox - topbox + 1);
    
    // Draw the scroll indicator
    _u8g2->drawBox(leftbox + 1, offset, scroll_bar_width - 2, bar_size);
    
    sendBuffer();
}

bool VFD_SH1106_Adapter::write(const char* str) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    _u8g2->setCursor(_cursorX, _cursorY);
    _u8g2->drawStr(_cursorX, _cursorY, str);
    sendBuffer();
    return true;
}

bool VFD_SH1106_Adapter::write(string str) {
    return write(str.c_str());
}

bool VFD_SH1106_Adapter::writePacket(const uint8_t *data, size_t len, useconds_t waitusec) {
    // This is a simplified implementation that doesn't handle all the special commands
    // from the original VFD class. For complex commands, we'd need to parse and interpret them.
    
    if (!_isSetup || _u8g2 == nullptr || len == 0)
        return false;
    
    // Check for special commands
    if (len > 0) {
        switch (data[0]) {
            case VFD_OUTLINE:
                if (len >= 5) {
                    // Draw a frame
                    _u8g2->drawFrame(data[1], data[2], data[3] - data[1] + 1, data[4] - data[2] + 1);
                    sendBuffer();
                    return true;
                }
                break;
                
            case VFD_CLEAR_AREA:
                if (len >= 5) {
                    // Clear an area by drawing a filled box with background color
                    _u8g2->setDrawColor(0);
                    _u8g2->drawBox(data[1], data[2], data[3] - data[1] + 1, data[4] - data[2] + 1);
                    _u8g2->setDrawColor(1);
                    sendBuffer();
                    return true;
                }
                break;
                
            case VFD_SET_AREA:
                if (len >= 5) {
                    // Fill an area
                    _u8g2->drawBox(data[1], data[2], data[3] - data[1] + 1, data[4] - data[2] + 1);
                    sendBuffer();
                    return true;
                }
                break;
                
            case VFD_SET_CURSOR:
                if (len >= 3) {
                    setCursor(data[1], data[2]);
                    return true;
                }
                break;
                
            // Handle other special commands as needed
        }
    }
    
    // If not a special command, treat as text
    char* text = new char[len + 1];
    memcpy(text, data, len);
    text[len] = '\0';
    
    bool result = write(text);
    delete[] text;
    
    return result;
}

bool VFD_SH1106_Adapter::printPacket(const char *fmt, ...) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    char* s;
    va_list args;
    va_start(args, fmt);
    vasprintf(&s, fmt, args);
    
    bool success = write(s);
    free(s);
    va_end(args);
    
    return success;
}

bool VFD_SH1106_Adapter::printLines(uint8_t y, uint8_t step, vector<string> lines,
                                    uint8_t firstLine, uint8_t maxLines, 
                                    font_t font) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    auto lineCount = lines.size();
    
    setFont(font);
    
    if (maxLines >= lineCount) {
        // Ignore the offset and draw all lines
        for (int i = 0; i < lineCount; i++) {
            string str = lines[i];
            if (!str.empty()) {
                setCursor(0, y);
                bool success = printPacket("%s", str.c_str());
                if (!success) return false;
            }
            y += step;
        }
    } else {
        // This text needs to be scrolled
        auto maxFirstLine = lineCount - maxLines;
        if (firstLine > maxFirstLine) firstLine = maxFirstLine;
        
        auto count = lineCount - firstLine;
        if (count > maxLines) count = maxLines;
        
        for (auto i = firstLine; i < firstLine + count; i++) {
            string str = lines[i];
            
            // Calculate available width considering scroll bar
            uint8_t rightbox = width() - 1 - scroll_bar_width;
            
            // Clear the line area
            _u8g2->setDrawColor(0);
            _u8g2->drawBox(0, y - step + 1, rightbox, step - 1);
            _u8g2->setDrawColor(1);
            
            setCursor(0, y);
            bool success = printPacket("%s", str.c_str());
            if (!success) return false;
            
            y += step;
        }
    }
    
    return true;
}

bool VFD_SH1106_Adapter::printRows(uint8_t y, uint8_t step, vector<vector<string>> columns,
                                   uint8_t firstLine, uint8_t maxLines, uint8_t col1_start,
                                   font_t font) {
    if (!_isSetup || _u8g2 == nullptr)
        return false;
    
    auto lineCount = columns.size();
    
    setFont(font);
    
    // Calculate column widths
    uint8_t longest_col1_pixel_width = 0;
    
    for (auto &row : columns) {
        if (row.size() > 0 && !row[0].empty()) {
            // Estimate width based on character count and font
            int charWidth;
            switch (font) {
                case FONT_MINI: charWidth = 5; break;
                case FONT_5x7: charWidth = 6; break;
                case FONT_10x14: charWidth = 10; break;
                default: charWidth = 5;
            }
            
            uint8_t length = row[0].length() * charWidth;
            if (length > longest_col1_pixel_width)
                longest_col1_pixel_width = length;
        }
    }
    
    uint8_t col2_start = col1_start + longest_col1_pixel_width + 2;
    
    if (maxLines >= lineCount) {
        // Ignore the offset and draw all
        for (int i = 0; i < lineCount; i++) {
            vector<string> row = columns[i];
            string str = row[0];
            string col2 = "";
            if (row.size() > 1) col2 = row[1];
            
            setCursor(col1_start, y);
            bool success = printPacket("%s", str.c_str());
            if (!success) return false;
            
            setCursor(col2_start, y);
            success = printPacket("%s", col2.c_str());
            if (!success) return false;
            
            y += step;
        }
    } else {
        // This text needs to be scrolled
        auto maxFirstLine = lineCount - maxLines;
        if (firstLine > maxFirstLine) firstLine = maxFirstLine;
        
        auto count = lineCount - firstLine;
        if (count > maxLines) count = maxLines;
        
        for (auto i = firstLine; i < firstLine + count; i++) {
            vector<string> row = columns[i];
            string str = row[0];
            string col2 = "";
            if (row.size() > 1) col2 = row[1];
            
            // Clear the line area
            _u8g2->setDrawColor(0);
            _u8g2->drawBox(col1_start, y - step + 1, width() - col1_start - scroll_bar_width, step - 1);
            _u8g2->setDrawColor(1);
            
            setCursor(col1_start, y);
            bool success = printPacket("%s", str.c_str());
            
            setCursor(col2_start, y);
            if (success && !col2.empty()) {
                success = printPacket("%s", col2.c_str());
            }
            
            if (!success) return false;
            
            y += step;
        }
    }
    
    return true;
}

void VFD_SH1106_Adapter::sendBuffer() {
    if (_isSetup && _u8g2 != nullptr) {
        _u8g2->sendBuffer();
    }
}
